File: apps/ios/Wewed/Services/WeddingDayGateAwareRepository.swift
import Foundation

/// Preserves all existing native surfaces while routing Gate QR admission through Wedding Day.
public actor WeddingDayGateAwareRepository: WeddingRepositoryProtocol {
    private let base: WeddingRepositoryProtocol
    private let gate: WeddingDayGateOperations

    public init(base: WeddingRepositoryProtocol, gate: WeddingDayGateOperations) {
        self.base = base
        self.gate = gate
    }

    public func availableWeddingIds() async throws -> [String] { try await base.availableWeddingIds() }
    public func resolveGuestIdentity(token: String) async throws -> GuestIdentity? {
        try await base.resolveGuestIdentity(token: token)
    }
    public func getWedding(weddingId: String) async throws -> Wedding { try await base.getWedding(weddingId: weddingId) }
    public func getTasks(weddingId: String) async throws -> [PlannerTask] { try await base.getTasks(weddingId: weddingId) }
    public func createTask(weddingId: String, title: String, priority: TaskPriority, category: String) async throws -> PlannerTask {
        try await base.createTask(weddingId: weddingId, title: title, priority: priority, category: category)
    }
    public func toggleTask(weddingId: String, taskId: String) async throws -> PlannerTask {
        try await base.toggleTask(weddingId: weddingId, taskId: taskId)
    }
    public func getGuests(weddingId: String) async throws -> [Guest] { try await base.getGuests(weddingId: weddingId) }
    public func getBudget(weddingId: String) async throws -> BudgetSummary { try await base.getBudget(weddingId: weddingId) }
    public func getWeddingPass(token: String) async throws -> WeddingPass { try await base.getWeddingPass(token: token) }
    public func searchGuests(weddingId: String, query: String) async throws -> [Guest] {
        try await base.searchGuests(weddingId: weddingId, query: query)
    }
    public func getAuditRecords(weddingId: String) async throws -> [CheckInAuditRecord] {
        try await base.getAuditRecords(weddingId: weddingId)
    }
    public func getVendors(weddingId: String) async throws -> [VendorPresence] { try await base.getVendors(weddingId: weddingId) }
    public func updateVendorState(weddingId: String, id: String, state: VendorPresenceState) async throws -> VendorPresence {
        try await base.updateVendorState(weddingId: weddingId, id: id, state: state)
    }
    public func getAnnouncements(weddingId: String) async throws -> [WeddingAnnouncement] {
        try await base.getAnnouncements(weddingId: weddingId)
    }
    public func postAnnouncement(weddingId: String, title: String, message: String, urgency: AnnouncementUrgency) async throws -> WeddingAnnouncement {
        try await base.postAnnouncement(weddingId: weddingId, title: title, message: message, urgency: urgency)
    }
    public func resolveInvitation(weddingSlug: String, token: String) async throws -> InvitationContext {
        try await base.resolveInvitation(weddingSlug: weddingSlug, token: token)
    }
    public func confirmRsvp(weddingSlug: String, token: String, attending: Bool) async throws -> WeddingPass {
        try await base.confirmRsvp(weddingSlug: weddingSlug, token: token, attending: attending)
    }

    /// Gate admission stays routed through Wedding Day, but still inside the wedding scope.
    public func checkInGuest(weddingId: String, qrPayload: String, count: Int, usherId: String) async throws -> CheckInVerificationResult {
        let available = try await base.availableWeddingIds()
        guard available.contains(weddingId) else {
            throw WeddingScopeMismatch(requestedWeddingId: weddingId, availableWeddingIds: available)
        }
        guard weddingId == gate.gateContext.weddingId else {
            throw WeddingScopeMismatch(
                requestedWeddingId: weddingId,
                availableWeddingIds: [gate.gateContext.weddingId]
            )
        }
        // usherId exists only on the legacy repository protocol; it is intentionally ignored.
        return try await gate.checkIn(qrPayload: qrPayload, count: count)
    }
}

File: apps/android/app/src/main/java/pro/wewed/app/services/WeddingDayGateAwareRepository.kt
package pro.wewed.app.services

import pro.wewed.app.models.*

/** Preserves all existing native surfaces while routing Gate QR admission through Wedding Day. */
class WeddingDayGateAwareRepository(
    private val base: WeddingRepository,
    private val gate: WeddingDayGateOperations
) : WeddingRepository {
    override suspend fun availableWeddingIds(): List<String> = base.availableWeddingIds()
    override suspend fun getWedding(weddingId: String): Wedding = base.getWedding(weddingId)
    override suspend fun getTasks(weddingId: String): List<PlannerTask> = base.getTasks(weddingId)
    override suspend fun createTask(weddingId: String, title: String, priority: TaskPriority, category: String): PlannerTask =
        base.createTask(weddingId, title, priority, category)
    override suspend fun toggleTask(weddingId: String, taskId: String): PlannerTask = base.toggleTask(weddingId, taskId)
    override suspend fun getGuests(weddingId: String): List<Guest> = base.getGuests(weddingId)
    override suspend fun getBudget(weddingId: String): BudgetSummary = base.getBudget(weddingId)
    override suspend fun getWeddingPass(token: String): WeddingPass = base.getWeddingPass(token)
    override suspend fun searchGuests(weddingId: String, query: String): List<Guest> = base.searchGuests(weddingId, query)
    override suspend fun getAuditRecords(weddingId: String): List<CheckInAuditRecord> = base.getAuditRecords(weddingId)
    override suspend fun getVendors(weddingId: String): List<VendorPresence> = base.getVendors(weddingId)
    override suspend fun updateVendorState(weddingId: String, id: String, state: VendorPresenceState): VendorPresence =
        base.updateVendorState(weddingId, id, state)
    override suspend fun getAnnouncements(weddingId: String): List<WeddingAnnouncement> = base.getAnnouncements(weddingId)
    override suspend fun postAnnouncement(
        weddingId: String,
        title: String,
        message: String,
        urgency: AnnouncementUrgency
    ): WeddingAnnouncement = base.postAnnouncement(weddingId, title, message, urgency)
    override suspend fun resolveInvitation(weddingSlug: String, token: String): InvitationContext =
        base.resolveInvitation(weddingSlug, token)
    override suspend fun confirmRsvp(weddingSlug: String, token: String, attending: Boolean): WeddingPass =
        base.confirmRsvp(weddingSlug, token, attending)

    /** Gate admission stays routed through Wedding Day, but still inside the wedding scope. */
    override suspend fun checkInGuest(
        weddingId: String,
        qrPayload: String,
        count: Int,
        usherId: String
    ): CheckInVerificationResult {
        val available = base.availableWeddingIds()
        if (weddingId !in available) throw WeddingScopeMismatch(weddingId, available)
        return gate.checkIn(qrPayload, count, usherId)
    }
}
